arvo/hoon.hoon

```diff
@@ -2452,10 +2452,15 @@
     $(a r.a, +<+.b $(a l.a, +<+.b (b n.a +<+.b)))
   ::
   +-  run                                               ::  apply gate to values
-    |*  {b/gate c/*}
-    |-
-    ?~  a  c
-    $(a r.a, c [(b n.a) $(a l.a)])
+    ~/  %run
+    |*  b/gate
+    =|  c/(set _?>(?=(^ a) (b n.a)))
+    |-  ?~  a  c
+    =+  ^=  c
+      (~(put in c) (b n.a))
+    =+  ^=  c
+      $(a l.a, c c)
+    $(a r.a, c c)
   ::
```

Comment:
Do what looks nicer to you, but whatever you do, avoid closing bracket pileups by using tall forms. (The type incantation in the `=|` is a piece of cosmic insight, so that should stay on one line to avoid taking up the attention of the uninitiated.)